### PR TITLE
Make orderBy processing in WindowAgg a bit less imperative

### PR DESCRIPTION
**Pull request by mfussenegger**
## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

**Comment by mfussenegger** [comment]
@andreidan this is what I had in mind in a prior PR. Let me know what you think

File: sql/src/main/java/io/crate/analyze/OrderBy.java [hunk 1]
```diff
@@ -53,6 +69,55 @@ public OrderBy(List<Symbol> orderBySymbols, boolean[] reverseFlags, Boolean[] nu
...
+     * Creates a new OrderBy with the other symbols prepended (or returns this if symbols are empty).
+     * Symbols are de-duplicated to some degree, e.g.
+     *
+     * <pre>
+     *     ORDER BY x y `prependUnique` x z will result in x z y; without duplicating x
+     * </pre>
+     *
+     * The defaults for reverseFlags and nullsFirst are used (asc, undefined)
+     */
+    public OrderBy prependUnique(Collection<? extends Symbol> symbols) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/OrderBy.java]: Maybe `prependNonDuplicates` would be clearer. Thoughts?

Comment (kovrus) [reply sql/src/main/java/io/crate/analyze/OrderBy.java]: imho, `prependUnique` is ok, clear and concise.

File: sql/src/main/java/io/crate/analyze/OrderBy.java [hunk 2]
```diff
@@ -53,6 +69,55 @@ public OrderBy(List<Symbol> orderBySymbols, boolean[] reverseFlags, Boolean[] nu
...
+        if (symbols.isEmpty()) {
+            return this;
+        }
+        ArrayList<Symbol> newOrderBySymbols = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/analyze/OrderBy.java]: Didn't do it since it may be too large, but I guess you're right in that it would likely still be better than the default. Will add it.

File: sql/src/test/java/io/crate/planner/operators/WindowAggTest.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Test
+    public void testOrderByIsMergedWithPartitionByWithPartialColumnOverlap() {
+        OrderBy orderBy = WindowAgg.createOrderByInclPartitionBy(wd("avg(x) OVER (PARTITION BY x, y ORDER BY x)"));
+        assertThat(orderBy, notNullValue());
+        assertThat(orderBy.orderBySymbols(), contains(isField("x"), isField("y")));
+    }
+
+    @Test
+    public void testOrderByIsMergedWithPartitionByWithPartialColumnOverlapButWrongOrder() {
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/test/java/io/crate/planner/operators/WindowAggTest.java]: The "wrong order" is referring to the partial overlap. The `x` in `ORDER BY x`  isn't at the first position in `PARTITION BY y, x`

File: sql/src/main/java/io/crate/analyze/OrderBy.java [hunk 4]
```diff
@@ -53,6 +69,55 @@ public OrderBy(List<Symbol> orderBySymbols, boolean[] reverseFlags, Boolean[] nu
...
+        while (xsToPrepend.hasNext()) {
+            Symbol toPrepend = xsToPrepend.next();
+            if (nextOrderBy == null && orderBySymbols.hasNext()) {
+                nextOrderBy = orderBySymbols.next();
+            }
+            if (toPrepend.equals(nextOrderBy)) {
+                newOrderBySymbols.add(nextOrderBy);
+                newReverseFlags.add(reverseFlags[orderBySymbols.previousIndex()]);
+                newNullsFirst.add(nullsFirst[orderBySymbols.previousIndex()]);
+                nextOrderBy = null;
```
[the remark this answers is not in the document]
Comment (andreidan) [reply sql/src/main/java/io/crate/analyze/OrderBy.java]: on second thought, this null "flagging" is a bit misleading and hard to follow. should we set the `nextOrderBy` value to the next value here instead of above? 

```
if (orderBySymbols.hasNext()) {
    nextOrderBy = orderBySymbols.next();
} else {
    nextOrderBy = null;
}

[the remark this answers is not in the document]
Comment (andreidan) [reply sql/src/main/java/io/crate/analyze/OrderBy.java]: @mfussenegger what do you think of this ^ ?

File: sql/src/test/java/io/crate/planner/operators/WindowAggTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreidan) [reply sql/src/test/java/io/crate/planner/operators/WindowAggTest.java]: yes, sorry, I got there in the end :) Do you think "reverse order" makes a bit more sense? (it's less confusing to me, as I was looking for something to be "wrong" in the query)

File: sql/src/main/java/io/crate/analyze/OrderBy.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/analyze/OrderBy.java]: @andreidan pushed a fixup
